### PR TITLE
DX12: Align copies b/w textures and buffers when `D3D12_FEATURE_DATA_D3D12_OPTIONS13.UnrestrictedBufferTextureCopyPitchSupported` is `false`

### LINKED ISSUE #4150: [d3d12] Support padded `rows_per_image` for 3D texture writes
See https://github.com/gfx-rs/wgpu/pull/3992#discussion_r1328532200.

Impact of this not working is limited but as a user I imagine it would be tough figuring out why things don't behave as expected.

### LINKED ISSUE #5285: [d3d12] Panic in end_encoding
CTS test: `cts/webgpu/api/operation/command_buffer/image_copy/cts.https.html?q=webgpu:api,operation,command_buffer,image_copy:undefined_params:*`

```
10  xul.dll!core::panicking::panic() [panicking.rs:07dca489ac2d933c78d3c5158e3f43beefeb02ce : 144 + 0x7]
     rbx = 0x0000018e79454b00    rbp = 0x0000006fa2f3a6c0
     rsp = 0x0000006fa2f3a630    r12 = 0x0000006fa2f3c8d0
     r13 = 0x0000018e79358c00    r14 = 0x0000006fa2f3d9d0
     r15 = 0x0000018e79454000    rip = 0x00007ff895816e42
    Found by: call frame info
11  xul.dll!enum2$<core::option::Option<d3d12::com::ComPtr<winapi::um::d3d12::ID3D12GraphicsCommandList> > >::take() [option.rs : 931]
    Found by: inlining
12  xul.dll!wgpu_hal::dx12::command::impl$3::end_encoding(wgpu_hal::dx12::CommandEncoder*) [command.rs:94c7d956509d4818aa4580c84b48e16a6933627b : 295 + 0x17]
     rbx = 0x0000018e79454b00    rbp = 0x0000006fa2f3a6c0
     rsp = 0x0000006fa2f3a6a0    r12 = 0x0000006fa2f3c8d0
     r13 = 0x0000018e79358c00    r14 = 0x0000006fa2f3d9d0
     r15 = 0x0000018e79454000    rip = 0x00007ff894b858b4
    Found by: call frame info
13  xul.dll!wgpu_core::device::queue::PendingWrites<wgpu_hal::dx12::Api>::pre_submit() [queue.rs:94c7d956509d4818aa4580c84b48e16a6933627b : 236]
    Found by: inlining
14  xul.dll!wgpu_core::global::Global::queue_submit<wgpu_hal::dx12::Api>(wgpu_core::id::Id<enum2$<wgpu_core::id::markers::Queue> >, ref$<slice2$<wgpu_core::id::Id<enum2$<wgpu_core::id::markers::CommandBuffer> > > >) [queue.rs:94c7d956509d4818aa4580c84b48e16a6933627b : 1466 + 0x20c]
     rbx = 0x0000018e79454b00    rsi = 0x0000000000000000
     rbp = 0x0000006fa2f3a760    rsp = 0x0000006fa2f3a6e0
     r12 = 0x0000006fa2f3c8d0    r13 = 0x0000018e79358c00
     r14 = 0x0000006fa2f3d9d0    r15 = 0x0000018e79454000
     rip = 0x00007ff8936bee2f
    Found by: call frame info
```

log:

```
[WARN  wgpu_core::instance] Missing downlevel flags: DownlevelFlags(VERTEX_AND_INSTANCE_INDEX_RESPECTS_RESPECTIVE_FIRST_VALUE_IN_INDIRECT_DRAW)
    The underlying API or device in use does not support enough features to be a fully compliant implementation of WebGPU. A subset of the features can still be used. If you are running this program on native and not in a browser and wish to limit the features you use to the supported subset, call Adapter::downlevel_properties or Device::downlevel_properties to get a listing of the features the current platform supports.
[WARN  wgpu_core::instance] DownlevelCapabilities {
        flags: DownlevelFlags(
            COMPUTE_SHADERS | FRAGMENT_WRITABLE_STORAGE | INDIRECT_EXECUTION | BASE_VERTEX | READ_ONLY_DEPTH_STENCIL | NON_POWER_OF_TWO_MIPMAPPED_TEXTURES | CUBE_ARRAY_TEXTURES | COMPARISON_SAMPLERS | INDEPENDENT_BLEND | VERTEX_STORAGE | ANISOTROPIC_FILTERING | FRAGMENT_STORAGE | MULTISAMPLED_SHADING | DEPTH_TEXTURE_AND_BUFFER_COPIES | WEBGPU_TEXTURE_FORMAT_SUPPORT | BUFFER_BINDINGS_NOT_16_BYTE_ALIGNED | UNRESTRICTED_INDEX_BUFFER | FULL_DRAW_INDEX_UINT32 | DEPTH_BIAS_CLAMP | VIEW_FORMATS | UNRESTRICTED_EXTERNAL_TEXTURE_COPIES | SURFACE_VIEW_FORMATS | NONBLOCKING_QUERY_RESOLVE,
        ),
        limits: DownlevelLimits,
        shader_model: Sm5,
    }
[ERROR wgpu_hal::auxil::dxgi::exception] ID3D12CommandList::CopyTextureRegion: D3D12_PLACED_SUBRESOURCE_FOOTPRINT::Offset must be a multiple of 512 (aka. D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT) when UnrestrictedBufferTextureCopyPitchSupported is false. Offset is 768. [ RESOURCE_MANIPULATION ERROR #864: COPYTEXTUREREGION_INVALIDSRCOFFSET]
[ERROR wgpu_hal::auxil::dxgi::result] GraphicsCommandList::close failed: invalid argument
```

### PR DESCRIPTION
**Pull request by ErichDonGubler**
**Connections**

- Likely resolves #5285.
- A case we need to handle for #4150.

**Testing**

- @teoxoy and I proved on some custom DLLs copied from an old Windows machine that this works, both on older WARP adapters and when forcefully done in updated DX12 runtimes. So, we're definitely not breaking anything! It's hard to set up an environment to specifically test this, though.

**Squash or Rebase?**

rebase plz

**Checklist**

- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry.

### CODE REVIEW

**Comment by ErichDonGubler** [comment]
I'm...honestly not sure why this is [failing on the `wgpu_gpu::regression::issue_6827::test_scatter` test entry in CI](https://github.com/gfx-rs/wgpu/actions/runs/15227402149/job/42831119331?pr=7721#step:16:1860). I think I need some help. Marking as ready for review; perhaps @teoxoy or @cwfitzgerald could help me out?

File: wgpu-hal/src/dx12/mod.rs [hunk 1]
```diff
@@ -1413,8 +1432,11 @@ impl crate::Queue for Queue {
     ) -> Result<(), crate::DeviceError> {
         let mut temp_lists = self.temp_lists.lock();
         temp_lists.clear();
+        let mut intermediate_copy_bufs = self.intermediate_copy_bufs.lock();
         for cmd_buf in command_buffers {
             temp_lists.push(Some(cmd_buf.raw.clone().into()));
+            intermediate_copy_bufs.push(Arc::clone(&cmd_buf.intermediate_copy_bufs));
+            // TODO: When to clean accumulated copy buffers up?
```
Comment (ErichDonGubler) [inline wgpu-hal/src/dx12/mod.rs]: I'm thinking that the best time to clean these up might be during a `Poll`, where we can guarantee that temporary buffers are no longer necessary. Validation, please? 🥺

Comment (teoxoy) [reply wgpu-hal/src/dx12/mod.rs]: We can store them in the `CommandEncoder` and they will be cleaned up on `CommandEncoder::Drop`, `wgpu-core` holds HAL command encoders alive until the end of the submission.

**Comment by teoxoy** [comment]
The test is failing because of:

> Panic: GraphicsCommandList::close failed: The parameter is incorrect. (0x80070057)

I'd run the test locally with the validation layers enabled to see if there is more info. I've seen the error come up in cases where we didn't uphold some D3D12 invariant while encoding.

**Comment by ErichDonGubler** [comment]
@teoxoy and I spent some time enabling debug layers, and even forcing the same version of WARP that CI is using in the test that's failing with only a WARP adapter. We're not able to reproduce this locally on either of our machines. 😓

**Comment by cwfitzgerald** [comment]
I wonder if this is a windows 10 runtime thing. There's something going wrong in the runtime on Win10, but Win11's runtime deals with it fine. We would need to investigate setting up the Agility SDK

**Comment by ErichDonGubler** [comment]
Got the CI issue resolved! I was definitely doing something wrong. 👉🏻👈🏻 From [community chat](https://matrix.to/#/!FZyQrssSlHEZqrYcOb:matrix.org/$ZOuynR--KoUigX1P1oi92tHd_QP3r5tX8QhNC1eubjE?via=matrix.org&via=mozilla.org&via=tchncs.de): 


> @ErichDonGubler:

> I just paired with Teo, and we actually FIGGERED IT OUT!
>
> @teoxoy and I sideloaded a bunch of DLLs from an ancient Windows 10 computer I had, which happened to have a WARP adapter that did not support low-alignment buffer-to-texture copies.
> 
> ...turns out I accidentally swapped the alignment requirement branches. 😅
> 
> We had to work around:
>
> - Not having the right DLLs (because new versions have lower alignment needs) and stealing them from an old machine I happened to have around.
> - Crashing when trying to compile indirect buffer validation shaders (because old versions?)
> - Feature unification working against us when trying to use WGPU_VALIDATION_INDIRECT_CALL=0
> - Erich being dumb and necessitating debugging in the first place
>
> It was an hours-long adventure, lemmetellyawhat.

> @teoxoy:
>
> It was fun :) 

> @cwfitzgerald:
>
> oh boy!
>
> that sounds like an adventure indeed

File: wgpu-hal/src/dx12/mod.rs (hunk 1 above)
Comment (ErichDonGubler) [reply wgpu-hal/src/dx12/mod.rs]: Alright, aididit. That was easy! 😀

Comment (teoxoy) [reply wgpu-hal/src/dx12/mod.rs]: > We can store them in the `CommandEncoder` and they will be cleaned up on `CommandEncoder::Drop`, `wgpu-core` holds HAL command encoders alive until the end of the submission.

Actually this is not true, for the same reason https://github.com/gfx-rs/wgpu/issues/7847 is an issue. wgpu-core will reuse command encoders and they will only be dropped when the device is.

I think we need a reset method on the encoder (@cwfitzgerald was also asking about this in matrix).

Comment (ErichDonGubler) [reply wgpu-hal/src/dx12/mod.rs]: Could we maybe add these buffers to `CommandBuffer`s, which would then be cleared out on `Device::maintain`, while we wait?

Comment (ErichDonGubler) [reply wgpu-hal/src/dx12/mod.rs]: Ah, finally catching up with <https://github.com/gfx-rs/wgpu/issues/7847#issuecomment-3024754825>, sory! I guess the premise of that issue is that `CommandBuffer`s need to recycle resources, too, otherwise they accumulate. So…sigh.

Comment (cwfitzgerald) [reply wgpu-hal/src/dx12/mod.rs]: @teoxoy wouldn't the [`reset_all`](https://wgpu.rs/doc/wgpu_hal/trait.CommandEncoder.html#tymethod.reset_all) entry point already be enough?

Comment (teoxoy) [reply wgpu-hal/src/dx12/mod.rs]: Oh yes, `reset_all` would suffice, it gets called in the `Drop` impl of `InnerCommandEncoder` just before we mark the encoder as ready to be reused.

Comment (ErichDonGubler) [reply wgpu-hal/src/dx12/mod.rs]: I implemented clearing of temporary copy buffers in fd850ff7. WDYT?

Comment (cwfitzgerald) [reply wgpu-hal/src/dx12/mod.rs]: LGTM

File: wgpu-hal/src/dx12/command.rs [hunk 2]
```diff
@@ -673,6 +781,39 @@ impl crate::CommandEncoder for super::CommandEncoder {
...
+                let orig_offset = r.buffer_layout.offset;
+                let Some((intermediate_to_dst_region, src)) = (unsafe {
+                    self.buf_tex_intermediate(
+                        r,
+                        src.format,
+                        |this, buf, size, intermediate_region| {
+                            copy_aligned(this, src, buf, intermediate_region);
+                            crate::BufferCopy {
+                                src_offset: orig_offset,
+                                dst_offset: 0,
```
Comment (teoxoy) [inline wgpu-hal/src/dx12/command.rs]: These seem swapped.

File: wgpu-hal/src/dx12/command.rs [hunk 3]
```diff
@@ -312,6 +312,78 @@ impl super::CommandEncoder {
...
+        let size = {
+            let copy_info = region.buffer_layout.get_buffer_texture_copy_info(
+                tex_fmt,
+                region.texture_base.aspect.map(),
+                &region.size.into(),
+            );
+            copy_info.unwrap().bytes_in_copy
+        };
+
+        let size = wgt::BufferSize::new(size)?;
```
Comment (teoxoy) [inline wgpu-hal/src/dx12/command.rs]: I think we can `unwrap` here, since it would simplify the code and 0 sized copies shouldn't appear at the hal level.

**Review by teoxoy — CHANGES_REQUESTED** [review]
Looks good overall but we should also enable some CTS tests like the one in https://github.com/gfx-rs/wgpu/issues/5285.

File: wgpu-hal/src/dx12/command.rs (hunk 3 above)
Comment (ErichDonGubler) [reply wgpu-hal/src/dx12/command.rs]: I feel comfortable testing that assumption in production! 😀 Simplified with a32789cc.

File: wgpu-hal/src/dx12/command.rs (hunk 2 above)
Comment (ErichDonGubler) [reply wgpu-hal/src/dx12/command.rs]: D'oh, you're totally right. Fixed with a3fba46b.

**Comment by ErichDonGubler** [comment]
Doing some branch pushes to check that `webgpu:api,operation,command_buffer,image_copy:undefined_params:*` fails before and after, and hopefully that it catches the bug @teoxoy found with region offsets being swapped on the tex-to-buf path. 🤞🏻

**Comment by ErichDonGubler** [comment]
Alright, so:

- <https://github.com/erichdongubler-mozilla/wgpu/actions/runs/18361970647/job/52307306029> is based on `trunk`, and has 4 failures, including some in `2d` copies and `3d` copies.
- <https://github.com/erichdongubler-mozilla/wgpu/actions/runs/18361971738/job/52307309175> is based on this PR's contents _before_ the fix Teo identified for offsets in tex-to-buf. There are still `2d` failures here, but fewer.
- <https://github.com/gfx-rs/wgpu/actions/runs/18362103668/job/52307694538?pr=7721> implements Teo's fix, and confirms that things get greener. Yay! Unfortunately, there are still breakages with `3d` copies, so this whole thing isn't passing yet, but I'd say this is a nice confirmation that we're fixing something.

I'll figure out the exact set of CTS tests to use tomorrow. If nothing else, I can use a (verbosely requested) subset of `webgpu:api,operation,command_buffer,image_copy:undefined_params:*`.

**Review by teoxoy — APPROVED** [review]

**Comment by ErichDonGubler** [comment]
I added baseline expectations (plus failures) for `webgpu:api,operation,command_buffer,image_copy:undefined_params:*` to our CTS list with 932567d242135677e779e06303f48bca8f46c007. I have amended the current tip of the PR to remove `fails-if(dx12)` from some of the `2d` variants of the added tests, thus proving that we fix something with this PR. Nice!

File: cts_runner/test.lst [hunk 4]
```diff
@@ -14,6 +14,15 @@ fails-if(vulkan) webgpu:api,operation,command_buffer,copyTextureToTexture:copy_d
 webgpu:api,operation,command_buffer,copyTextureToTexture:copy_depth_stencil:format="stencil8"
 // Fails with OOM in CI.
 fails-if(dx12) webgpu:api,operation,command_buffer,image_copy:offsets_and_sizes:*
+webgpu:api,operation,command_buffer,image_copy:undefined_params:initMethod="WriteTexture";checkMethod="FullCopyT2B";dimension="1d"
+webgpu:api,operation,command_buffer,image_copy:undefined_params:initMethod="WriteTexture";checkMethod="FullCopyT2B";dimension="2d"
+webgpu:api,operation,command_buffer,image_copy:undefined_params:initMethod="WriteTexture";checkMethod="FullCopyT2B";dimension="3d"
+webgpu:api,operation,command_buffer,image_copy:undefined_params:initMethod="WriteTexture";checkMethod="PartialCopyT2B";dimension="1d"
+fails-if(dx12) webgpu:api,operation,command_buffer,image_copy:undefined_params:initMethod="WriteTexture";checkMethod="PartialCopyT2B";dimension="2d"
```
Comment (ErichDonGubler) [inline cts_runner/test.lst:21]: It's odd that `webgpu:api,operation,command_buffer,image_copy:undefined_params:initMethod="WriteTexture";checkMethod="PartialCopyT2B";dimension="2d"` still fails after this PR's fix. I suspect there's more work to do here to make this correct.

Comment (ErichDonGubler) [reply cts_runner/test.lst:21]: Ah, it seems the full path that's failing is a single subtest exercising a 3rd dimension for the `"2d"` case: `copySize=[3,3,2];origin=["_undef_","_undef_","_undef_"];bytesPerRow=256;rowsPerImage=3`

Definitely follow-up work for something more edge-casey. Whew!

File: cts_runner/test.lst [hunk 5]
```diff
@@ -157,4 +166,4 @@ webgpu:shader,validation,statement,statement_behavior:invalid_statements:body="l
 webgpu:shader,validation,statement,statement_behavior:invalid_statements:body="loop8"
 webgpu:shader,validation,statement,statement_behavior:invalid_statements:body="switch1"
 //FAIL: 9 invalid_statements subtests due to https://github.com/gfx-rs/wgpu/issues/7733
-webgpu:api,validation,render_pipeline,vertex_state:many_attributes_overlapping:*
\ No newline at end of file
+webgpu:api,validation,render_pipeline,vertex_state:many_attributes_overlapping:*
```
Comment (ErichDonGubler) [inline cts_runner/test.lst:169]: @andyleiserson: FYI, adding a newline to the end of the file here.

File: CHANGELOG.md [hunk 6]
```diff
@@ -86,6 +86,10 @@ SamplerDescriptor {
 - WebGPU device requests now support the required limits `maxColorAttachments` and `maxColorAttachmentBytesPerSample`. By @evilpie in [#8328](https://github.com/gfx-rs/wgpu/pull/8328)
 - Reject binding indices that exceed `wgpu_types::Limits::max_bindings_per_bind_group` when deriving a bind group layout for a pipeline. By @jimblandy in [#8325](https://github.com/gfx-rs/wgpu/pull/8325).
 
+#### DX12
+
+- Align copies b/w textures and buffers via a single intermediate buffer per copy when `D3D12_FEATURE_DATA_D3D12_OPTIONS13.UnrestrictedBufferTextureCopyPitchSupported` is `false`. By @ErichDonGubler in [#7721](https://github.com/gfx-rs/wgpu/pull/7721).
+
```
Comment (ErichDonGubler) [inline CHANGELOG.md:92]: onuts, I almost forgot to move this out of the previous release section.

@cwfitzgerald: I'm strongly considering experimenting with adding some CI that fails a PR that adds things to the `CHANGELOG` outside of the `Unreleased` section, unless opted out by a label. Thoughts?

Comment (ErichDonGubler) [reply CHANGELOG.md:92]: I guess we don't have to make it a required check, either. It can simply be a signal that something might be wrong, at first.

**Comment by torokati44** [comment]
Oh, is a backport of this to the 25/26/27 series, and corresponding patch releases feasible? 👀 It certainly would be appreciated IMHO!

**Comment by ErichDonGubler** [comment]
@torokati44: Two things:

1. For optional fixes like this, we generally expect the community to handle backporting effort (which should be light here, since you can omit the test coverage). So, if you're willing, then we look forward to a PR from you against `v26` and `v27`. 😄 
2. In our maintainers' meeting today, we know we're willing to field backports to v26 (esp. for Bevy) and v27. Was there a reason you wanted to backport to v25, specifically?

**Comment by torokati44** [comment]
@ErichDonGubler:

1. Alright, I'll see what I can do w.r.t backporting PRs.

2. No, not really, anymore, since https://github.com/ruffle-rs/ruffle/pull/21730 got merged (but not yet https://github.com/ruffle-rs/ruffle/pull/21895)
